scapy_traceroute.py

```diff
@@ -0,0 +1,324 @@
+				pass			#No need to fill in, we already have None's there.
+			elif reply.haslayer(ICMPv6TimeExceeded) or (reply.haslayer(ICMP) and reply[ICMP].type == 11):	#Intermediate host (Type is time-exceeded)
+				hop_list[i] = explode_ip(ip_addr_obj(reply[address_layer].src))
+			elif reply.haslayer('ICMPv6 Destination Unreachable') or (reply.haslayer(ICMP) and reply[ICMP].type == 3):	#Reached target (Type is dest-unreach)
+				hop_list[i] = explode_ip(ip_addr_obj(reply[address_layer].src))
+				del hop_list[i+1:]	#Truncate any following nulls
+				break
+			else:				#Unknown
+				sys.stderr.write("Unknown reply type:\n")
+				reply.show()
+				break
+
+		#sys.stderr.write('\n')
+
+		try:
+			write_object(cache_file(tr_cache_dir, target), json.dumps(hop_list))
+		except:
+			pass
+
+		#Loop that truncates the list by one element, saves each sublist under the intermediate IP address (and mkdir_p that address too)
+		truncated_path_to_ip = list(hop_list)		#Make a shallow copy of the list so we don't affect the original
+		del truncated_path_to_ip[-1]			#Loop, dropping the rightmost entry each time.  Working back through the list of routers, save any that are actually routers for which we don't already have a path to that router.
+		while truncated_path_to_ip:
+			if truncated_path_to_ip[-1]:		#If not null, i.e., we have an actual router IP address:
+				router_ip = truncated_path_to_ip[-1]
+				if not os.path.exists(cache_file(tr_cache_dir, router_ip)):
+					#sys.stderr.write("____ writing router path:" + router_ip + ":" + str(truncated_path_to_ip)
+					#sys.stderr.flush
+					try:
+						write_object(cache_file(tr_cache_dir, router_ip), json.dumps(truncated_path_to_ip))
+					except:
+						pass
+
+			del truncated_path_to_ip[-1]
+
+	return hop_list
+
+
+
+scapy_traceroute_version = '0.2.5'
+per_packet_timeout_default = 1
+forced_interface_default = None
+ttl_default = 30
+traceroute_cache_dir_default = os.environ["HOME"] + '/.cache/scapy_traceroute/'
+
+if __name__ == '__main__':
+	import argparse
+
+	parser = argparse.ArgumentParser(description='scapy_traceroute version ' + str(scapy_traceroute_version))
+	parser.add_argument('-p', '--per_packet_timeout', help='Time to wait for a reply for a single packet, can be fractional (default: ' + str(per_packet_timeout_default) + ' ).', required=False, default=per_packet_timeout_default)
+	parser.add_argument('-f', '--forced_interface', help='Force packets through this interface (needed on macos, default: ' + str(forced_interface_default) + ' ).', required=False, default=forced_interface_default)
+	parser.add_argument('-t', '--ttl', help='Maximum number of hops to try (default: ' + str(ttl_default) + ')', required=False, default=ttl_default)
+	parser.add_argument('-c', '--cache_dir', help='Directory tree to hold cached traceroutes (default: ' + str(traceroute_cache_dir_default) + ' ).  Use None to not cache results.', required=False, default=traceroute_cache_dir_default)
```

Comment:
I don't believe that the string `None` passed on the cli will be properly cast to `None` in python. Instead, I believe it will come in as the string `"None"`. 

I haven't tested this, so please dismiss this if this is not an issue.